// Copyright (c) 2021 Huawei Device Co., Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_package = "ohos.devtools.datasources.transport.grpc.service";
option optimize_for = LITE_RUNTIME;

message SampleTimeStamp {
    uint64 tv_sec = 1;
    uint64 tv_nsec = 2;
}

message CpuCoreFrequency {
  int32 min_frequency_khz = 1;
  int32 max_frequency_khz = 2;
  int32 cur_frequency_khz = 3;
}

message CpuCoreUsageInfo {
    int32 cpu_core = 1;
    int64 prev_system_cpu_time_ms = 2;
    int64 prev_system_boot_time_ms = 3;
    int64 system_cpu_time_ms = 4;
    int64 system_boot_time_ms = 5;
    CpuCoreFrequency frequency = 6;
    bool is_little_core = 7;
}

message CpuUsageInfo {
    int64 prev_process_cpu_time_ms = 1;
    int64 prev_system_cpu_time_ms = 2;
    int64 prev_system_boot_time_ms = 3;
    int64 process_cpu_time_ms = 4;
    int64 system_cpu_time_ms = 5;
    int64 system_boot_time_ms = 6;
    repeated CpuCoreUsageInfo cores = 7;
    SampleTimeStamp timestamp = 8;
}

enum ThreadState {
    THREAD_UNSPECIFIED = 0;
    THREAD_RUNNING = 1;
    THREAD_SLEEPING = 2;
    THREAD_STOPPED = 3;
    THREAD_WAITING = 4;
}

message ThreadInfo {
    int32 tid = 1;
    string thread_name = 2;
    ThreadState thread_state = 3;
    int64 prev_thread_cpu_time_ms = 4;
    int64 thread_cpu_time_ms = 5;
    SampleTimeStamp timestamp = 6;
}

message CpuData {
    CpuUsageInfo cpu_usage_info = 1;
    repeated ThreadInfo thread_info = 2;
    int64 process_num = 3;
    double user_load = 4;
    double sys_load = 5;
    double total_load = 6;
}
